### ISSUE TITLE
How does the project scale when entities growing up?

### ISSUE THREAD

**Issue by sonphnt** [EduardoPires/EquinoxProject#127]
Hi, Firstly, Thanks for a great Repo. I have a question would like to ask.
I can see regarding to "Customer" entity. There are 2 projects that contain all relevant codes to "Customer"

Equinox.Application
--Services
--ViewModels

Equinox.Domain
--CommandHandler
--Commands
-- So on.....

If I have a new Entity "Product". I have to create appropriate files/codes in those 2 projects such as:
Equinox.Domain
--CommandHandler
----CustomerHandler
----CustomerCommandHandler
----ProductHandler
----ProductCommandHandler

the same with Commands/EventHandlers/Interfaces ......And at some point when I have a lot of entities. How do we easily manage files?

Thanks

**Comment by TeoDanilo** [comment]
Hello, since we don't know when will Eduardo answer, I'll give you a basic answer.

If you only want to create a Product Entity, you pretty much have only to create the appropriate files in all the layers, and adit some others (like EquinoxContext in Infra.Data layer for example). Don't forget to create a new Migration and update your database once you finish the code.

The the easy way to manage the files for me is to create aggregation roots, for example: In Domain layer, I would create a folder named 'Product' and put inside all the code and aggregated Entitys for products there, for example an Entity called Product.cs, then an Entity called ProductCatagory.cs and so on. ONLY ONE Repository for aggregation root though.

**Comment by sonphnt** [comment]
Hi @TeoDanilo . So you meant like this?

**Equinox.Domain**
**--Product**
----CommandHandler
------ProductCommandHandler.cs
----EventHandler
----Interfaces
**--Customer**
----CommandHandler
------CustomerCommandHandler.cs
----EventHandler
----Interfaces

**Comment by TeoDanilo** [comment]
Yes, @sonphnt, thats the idea.

I would do:
Equinox.Domain
**-Products**
   **---CommandHandler** (you could use the Commands folder to store this files if you like)
    _----ProductCommandHandler.cs_
   **---Commands**
    _----ProductCommands.cs_ (here you could create a separate file for eatch command class, like 'CreateNewProcuxtCommand.cs. I usualy put all the classes on the same file)
    _----CategorieCommands.cs_ (the same here)
   **---EventHandlers** (the same as CommandHandler)
   **---Events** (the same as commands)
   **---Interfaces**
    _----IProductRepository.cs_
   **---Validations**
    _----ProductValitadions.cs_
    _----CategorieValidations.cs_
  _--Product.cs_
  _--Categorie.cs_

**Comment by sonphnt** [comment]
@TeoDanilo Yes, I agree. With this template, we can easily copy/paste entirely Product folder and rename files for a new entity. The question is should we make each entity per folder like Product, Category. They are should be in 2 different folders instead of grouping into 1 folder?

**Comment by toiletspider** [comment]
@TeoDanilo do you perhaps know why the CustomerEventHandler.cs are in the Domain layer and not the Application layer? Looking at the CustomerEventHandler.cs, it is suppose to send emails, is that not something you would expect in Application.Services?

**Comment by TeoDanilo** [comment]
@toiletspider  I think basicaly: 

1 - CustomerEventHandler is not only to send emails, is been used by CustomerHistory (in Application layer) to store events. So it is a domain logic.
2 - Application layer is used by de MVC layer. Once you had another front end (like Angular, for example) you won't need it anymore, so you would have to remake the CustomerEventHandler in each front-end.  In Domain layer make it easier to use it from any front-end.

**Comment by TeoDanilo** [comment]
@toiletspider I think basicaly:

1 - CustomerEventHandler is not only to send emails, is used to store events. For example, is accessed by CustomerHistory (in Application layer) to show the updates done on the database and by who. So it is a domain logic.
2 - Application layer are used by de MVC layer. Once you had another front end (like Angular, for example) you won't need it anymore, so you would have to remake the CustomerEventHandler in each front-end. If it is on Domain layer make it easier to use it from any front-end.

**Comment by toiletspider** [comment]
@TeoDanilo thank you for the answer. I'm trying to get better understanding of DDD and CQRS, please bare with me:

![DDD Overview](https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/microservice-ddd-cqrs-patterns/media/image6.png)

The UI only works with the application layer(where the API are), so there will be no need to rewrite that code. See quote below for a comparison between MVC and DDD :
> Domain-driven design separates the model layer “M” of MVC into an application, domain and infrastructure layer. The infrastructure layer is used to retrieve and store data. The domain layer is where the business knowledge or expertise is. The application layer is responsible for coordinating the infrastructure and domain layers to make a useful application. Typically, it would use the infrastructure to obtain the data, consult the domain to see what should be done, and then use the infrastructure again to achieve the results
[DDD vs MVC](https://blog.fedecarg.com/2009/03/11/domain-driven-design-and-mvc-architectures/)

Concerning the above quote and [MS DDD](https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/microservice-ddd-cqrs-patterns/domain-events-design-implementation), the domain must not contain any reference to repositories/eventstore/services. Currently there are repositories in the Domain layer, it must only contain the Interfaces. It is my understanding that if Domain events are raised in the Domain Layer, they must be executed elsewhere. Example the writing to database should happen in the Infrastructure Layer (see diagram above). "The application layer is responsible for coordinating the infrastructure and domain layers to make a useful application", thus the event handler and command handler should be in the application layer to coordinate the writing to the database/sending emails.

The [Microsoft eshop](https://github.com/dotnet-architecture/eShopOnContainers/tree/23992ed3249cd257ac5fc42b23f2864b2ddf347b/src/Services/Ordering) which uses DDD and CQRS have the event handlers in Ordering.API/Application/DomainEventHandlers/

Have I misunderstood DDD and how these different layers interact or have it been implemented differently in this project than the references I have listed?

**Comment by chhinsras** [comment]
It seems confusing. Could any one help explaining me how to write this event sourcing.

I would like to use like _context.StoredEvent.Add(RegisteredCustomerEvent()));

Then it should trigger
